From 26edaced05d97b155d62145b8467d35358cc21b4 Mon Sep 17 00:00:00 2001
From: Oleksii Beketov <ol.beketov@samsung.com>
Date: Tue, 29 Jan 2019 16:47:10 +0200
Subject: [PATCH] Double free in oc_sec_free_certs_chain()

Fix for double free in oc_sec_free_certs_chain()

Change-Id: I0878113addb509db860dfc1e32469d6623066c59
Signed-off-by: Oleksii Beketov <ol.beketov@samsung.com>
---
 security/oc_tls.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/security/oc_tls.c b/security/oc_tls.c
index 0b7e5e59..98894724 100644
--- a/security/oc_tls.c
+++ b/security/oc_tls.c
@@ -494,12 +494,13 @@ oc_tls_add_peer(oc_endpoint_t *endpoint, int role)
 static void
 oc_sec_free_certs_chain(mbedtls_ssl_key_cert *kc)
 {
-  while (kc) {
-    mbedtls_x509_crt_free(kc->cert);
-    oc_mem_free(kc->cert);
-    mbedtls_pk_free(kc->key);
-    oc_mem_free(kc->key);
-    kc = kc->next;
+  mbedtls_ssl_key_cert *cur = kc, *next;
+  while(cur != NULL) {
+    next = cur->next;
+    mbedtls_x509_crt_free(cur->cert);
+    mbedtls_pk_free(cur->key);
+    oc_mem_free(cur);
+    cur = next;
   }
 }
 #if defined(OC_UNLOAD_CERT)
-- 
2.16.1.windows.1

